File: libUtil/src/main/java/org/waynezhou/libUtil/event/BaseEventGroup.java
package org.waynezhou.libUtil.event;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class BaseEventGroup<TEventGroup extends BaseEventGroup<TEventGroup>> {
    
    protected BaseEventGroup() {
    
    }
    
    @SuppressWarnings("unchecked")
    @NonNull
    public <TEventArgs> TEventGroup on(
      @NonNull EventGroupHolderSelector<TEventGroup, TEventArgs> selector,
      @NonNull EventListener<TEventArgs> listener
    ) {
        selector.select((TEventGroup) this).addOnListener(listener);
        return (TEventGroup) this;
    }
    
    @SuppressWarnings("unchecked")
    @NonNull
    public <TEventArgs> TEventGroup once(
      @NonNull EventGroupHolderSelector<TEventGroup, TEventArgs> selector,
      @NonNull EventListener<TEventArgs> listener
    ) {
        selector.select((TEventGroup) this).addOnceListener(listener);
        return (TEventGroup) this;
    }
    
    
    @SuppressWarnings("unchecked")
    @NonNull
    public <TEventArgs> EventHolder<TEventArgs>.ListenerToken on_(
      @NonNull EventGroupHolderSelector<TEventGroup, TEventArgs> selector,
      @NonNull EventListener<TEventArgs> listener
    ) {
        return selector.select((TEventGroup) this).addOnListener(listener);
    }
    
    @SuppressWarnings("unchecked")
    @NonNull
    public <TEventArgs> EventHolder<TEventArgs>.ListenerToken once_(
      @NonNull EventGroupHolderSelector<TEventGroup, TEventArgs> selector,
      @NonNull EventListener<TEventArgs> listener
    ) {
        return selector.select((TEventGroup) this).addOnceListener(listener);
    }
    
    public class Invoker {
        private final TEventGroup group;
        
        private Invoker(TEventGroup group) {
            this.group = group;
        }
        
        public <TEventArgs> void invoke(
          @NonNull EventGroupHolderSelector<TEventGroup, TEventArgs> selector,
          TEventArgs eventArgs,
          @NonNull Condition breakAction
        ) {
            selector.select(group).invoke(eventArgs, breakAction);
        }
    
        public <TEventArgs> void invoke(
          @NonNull EventGroupHolderSelector<TEventGroup, TEventArgs> selector,
          TEventArgs eventArgs
        ) {
            invoke(selector, eventArgs, ()->false);
        }
    }
    
    @Nullable
    private BaseEventGroup<TEventGroup>.Invoker invoker;
    
    @SuppressWarnings("unchecked")
    @NonNull
    protected Invoker getInvoker() {
        if (invoker == null) invoker = new Invoker((TEventGroup) this);
        return invoker;
    }
    
    @SuppressWarnings("unchecked")
    @NonNull
    public <TEventArgs> TEventGroup remove(
      @NonNull EventGroupHolderSelector<TEventGroup, TEventArgs> selector,
      @NonNull EventHolder<TEventArgs>.ListenerToken token
    ) {
        selector.select((TEventGroup) this).removeListener(token);
        return (TEventGroup) this;
    }
    
    @SuppressWarnings("unchecked")
    @NonNull
    public <TEventArgs> TEventGroup remove(
      @NonNull EventGroupHolderSelector<TEventGroup, TEventArgs> selector,
      @NonNull EventListener<TEventArgs> listener
    ) {
        selector.select((TEventGroup) this).removeListener(listener);
        return (TEventGroup) this;
    }
    
    @SuppressWarnings("unchecked")
    @NonNull
    public TEventGroup apply(@NonNull EventGroupApplier<TEventGroup> applier) {
        TEventGroup group = (TEventGroup) this;
        applier.apply(group);
        return group;
    }
    

    
    @FunctionalInterface
    public interface EventGroupApplier<TEventGroup extends BaseEventGroup<TEventGroup>> {
        void apply(@NonNull BaseEventGroup<TEventGroup> pack);
    }
    
    @FunctionalInterface
    public interface EventGroupHolderSelector<TEventGroup extends BaseEventGroup<TEventGroup>, TEventArgs> {
        @NonNull
        EventHolder<TEventArgs> select(@NonNull TEventGroup group);
    }

}

File: libUtilKt/src/main/java/org/waynezhou/libutilkt/event/BaseEventGroup.kt
package org.waynezhou.libutilkt.event

abstract class BaseEventGroup<TEventGroup : BaseEventGroup<TEventGroup>> protected constructor() {

    @Suppress("UNCHECKED_CAST")
    fun <TEventArgs> on(
        selector: (TEventGroup) -> EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): TEventGroup {
        selector(this as TEventGroup).addOnListener(listener)
        return this
    }

    @Suppress("UNCHECKED_CAST")
    fun <TEventArgs> once(
        selector: (TEventGroup) -> EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): TEventGroup {
        selector(this as TEventGroup).addOnceListener(listener)
        return this
    }

    @Suppress("UNCHECKED_CAST")
    fun <TEventArgs> tokenOn(
        selector: (TEventGroup) -> EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): EventHolder<TEventArgs>.ListenerToken {
        return selector(this as TEventGroup).addOnListener(listener)
    }

    @Suppress("UNCHECKED_CAST")
    fun <TEventArgs> tokenOnce(
        selector: (TEventGroup) -> EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): EventHolder<TEventArgs>.ListenerToken {
        return selector(this as TEventGroup).addOnceListener(listener)
    }

    inner class Invoker internal constructor(
        private val group: TEventGroup
    ) {
        fun <TEventArgs> invoke(
            selector: (TEventGroup) -> EventHolder<TEventArgs>,
            eventArgs: TEventArgs,
            breakAction: () -> Boolean = { false }
        ) {
            selector(group).invoke(eventArgs, breakAction)
        }
    }

    private var invoker: BaseEventGroup<TEventGroup>.Invoker? = null

    @Suppress("UNCHECKED_CAST")
    protected fun getInvoker(): BaseEventGroup<TEventGroup>.Invoker {
        if (invoker == null) invoker = Invoker(this as TEventGroup)
        return invoker!!
    }

    @Suppress("UNCHECKED_CAST")
     fun <TEventArgs> remove(
        selector: (TEventGroup) -> EventHolder<TEventArgs>,
        token: EventHolder<TEventArgs>.ListenerToken
    ): TEventGroup {
        selector(this as TEventGroup).removeListener(token)
        return this
    }

    @Suppress("UNCHECKED_CAST")
     fun <TEventArgs> remove(
        selector: (TEventGroup) -> EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): TEventGroup {
        selector(this as TEventGroup).removeListener(listener)
        return this
    }

}

class EventHolder<TEventArgs> {
    private inner class Listener(
        val isOnce: Boolean,
        val listener: EventListener<TEventArgs>
    )

    private val listeners: MutableList<Listener> = mutableListOf()
    internal fun addOnceListener(listener: EventListener<TEventArgs>): ListenerToken {
        val l = Listener(true, listener)
        val token = ListenerToken(l.hashCode())
        listeners.add(l)
        return token
    }

    internal fun addOnListener(listener: EventListener<TEventArgs>): ListenerToken {
        val l = Listener(false, listener)
        val token = ListenerToken(l.hashCode())
        listeners.add(l)
        return token
    }

    internal fun removeListener(listener: EventListener<TEventArgs>) {
        listeners.removeIf { x: Listener -> x.listener === listener }
    }

    internal fun removeListener(token: ListenerToken) {
        listeners.removeIf { x: Listener -> x.hashCode() == token.hashcode }
    }

    internal fun invoke(e: TEventArgs, breakAction: () -> Boolean) {
        val removedIdx: MutableList<Int> = mutableListOf()
        for ((idx, listener) in listeners.withIndex()) {
            listener.listener.invoke(e)
            if (listener.isOnce) {
                removedIdx.add(idx)
            }
            if (breakAction()) break
        }
        removedIdx.forEach { listeners.removeAt(it) }
    }

    inner class ListenerToken(val hashcode: Int)
}

fun interface EventListener<TEventArgs> {
    operator fun invoke(e: TEventArgs)
}